File: backend/app/models/education.py
"""
Educational content models for financial concepts and explanations.
"""

from sqlalchemy import Column, Integer, String, Text, DateTime, Boolean, ForeignKey, Table
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import declarative_base
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel
from enum import Enum

from app.core.database import Base


class ConceptType(str, Enum):
    """Types of financial concepts"""
    TECHNICAL_INDICATOR = "technical_indicator"
    FUNDAMENTAL_RATIO = "fundamental_ratio"
    MARKET_CONCEPT = "market_concept"
    TRADING_STRATEGY = "trading_strategy"
    RISK_MANAGEMENT = "risk_management"


class DifficultyLevel(str, Enum):
    """Difficulty levels for educational content"""
    BEGINNER = "beginner"
    INTERMEDIATE = "intermediate"
    ADVANCED = "advanced"


# Association table for concept relationships
concept_relationships = Table(
    'concept_relationships',
    Base.metadata,
    Column('parent_id', Integer, ForeignKey('educational_concepts.id'), primary_key=True),
    Column('child_id', Integer, ForeignKey('educational_concepts.id'), primary_key=True)
)


class EducationalConcept(Base):
    """Educational concept database model"""
    __tablename__ = "educational_concepts"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), unique=True, index=True, nullable=False)
    concept_type = Column(String(50), nullable=False)
    difficulty_level = Column(String(20), nullable=False)
    short_description = Column(String(500), nullable=False)
    detailed_explanation = Column(Text, nullable=False)
    practical_example = Column(Text, nullable=True)
    formula = Column(String(200), nullable=True)
    interpretation_guide = Column(Text, nullable=True)
    common_mistakes = Column(Text, nullable=True)
    keywords = Column(String(500), nullable=True)  # Comma-separated keywords
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    # Self-referential many-to-many relationship for related concepts
    related_concepts = relationship(
        "EducationalConcept",
        secondary=concept_relationships,
        primaryjoin=id == concept_relationships.c.parent_id,
        secondaryjoin=id == concept_relationships.c.child_id,
        back_populates="related_concepts"
    )

    # One-to-many relationship with learning paths
    learning_paths = relationship("LearningPath", back_populates="concepts")


class LearningPath(Base):
    """Learning path database model"""
    __tablename__ = "learning_paths"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(100), nullable=False)
    description = Column(Text, nullable=False)
    difficulty_level = Column(String(20), nullable=False)
    estimated_duration_minutes = Column(Integer, nullable=False)
    concept_id = Column(Integer, ForeignKey("educational_concepts.id"))
    order_index = Column(Integer, nullable=False)
    is_active = Column(Boolean, default=True)
    created_at = Column(DateTime, default=datetime.utcnow)

    # Relationship back to concept
    concepts = relationship("EducationalConcept", back_populates="learning_paths")


class UserLearningProgress(Base):
    """User learning progress tracking"""
    __tablename__ = "user_learning_progress"

    id = Column(Integer, primary_key=True, index=True)
    user_id = Column(Integer, ForeignKey("users.id"), nullable=False)
    concept_id = Column(Integer, ForeignKey("educational_concepts.id"), nullable=False)
    is_completed = Column(Boolean, default=False)
    completion_date = Column(DateTime, nullable=True)
    difficulty_rating = Column(Integer, nullable=True)  # 1-5 scale
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)


# Pydantic models for API responses
class EducationalConceptBase(BaseModel):
    """Base educational concept schema"""
    name: str
    concept_type: ConceptType
    difficulty_level: DifficultyLevel
    short_description: str
    detailed_explanation: str
    practical_example: Optional[str] = None
    formula: Optional[str] = None
    interpretation_guide: Optional[str] = None
    common_mistakes: Optional[str] = None
    keywords: Optional[str] = None


class EducationalConceptCreate(EducationalConceptBase):
    """Schema for creating educational concepts"""
    pass


class EducationalConceptResponse(EducationalConceptBase):
    """Schema for educational concept responses"""
    id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    related_concepts: List['EducationalConceptSummary'] = []

    class Config:
        from_attributes = True


class EducationalConceptSummary(BaseModel):
    """Summary schema for related concepts"""
    id: int
    name: str
    concept_type: ConceptType
    difficulty_level: DifficultyLevel
    short_description: str

    class Config:
        from_attributes = True


class LearningPathBase(BaseModel):
    """Base learning path schema"""
    name: str
    description: str
    difficulty_level: DifficultyLevel
    estimated_duration_minutes: int


class LearningPathCreate(LearningPathBase):
    """Schema for creating learning paths"""
    concept_ids: List[int]


class LearningPathResponse(LearningPathBase):
    """Schema for learning path responses"""
    id: int
    concepts: List[EducationalConceptSummary] = []
    is_active: bool
    created_at: datetime

    class Config:
        from_attributes = True


class ConceptExplanationRequest(BaseModel):
    """Request schema for concept explanations"""
    concept_name: str
    context: Optional[str] = None  # Context like stock symbol, analysis type
    user_level: Optional[DifficultyLevel] = DifficultyLevel.BEGINNER


class ConceptExplanationResponse(BaseModel):
    """Response schema for concept explanations"""
    concept: EducationalConceptResponse
    contextual_explanation: str
    related_suggestions: List[EducationalConceptSummary] = []
    next_learning_steps: List[str] = []


class UserLearningProgressBase(BaseModel):
    """Base user learning progress schema"""
    concept_id: int
    is_completed: bool
    difficulty_rating: Optional[int] = None


class UserLearningProgressCreate(UserLearningProgressBase):
    """Schema for creating user learning progress"""
    pass


class UserLearningProgressResponse(UserLearningProgressBase):
    """Schema for user learning progress responses"""
    id: int
    user_id: int
    concept: EducationalConceptSummary
    completion_date: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True

File: frontend/src/services/education.ts
/**
 * Educational service for managing learning content and progress
 */

export interface EducationalConcept {
  id: number;
  name: string;
  concept_type: string;
  difficulty_level: string;
  short_description: string;
  detailed_explanation: string;
  practical_example?: string;
  formula?: string;
  interpretation_guide?: string;
  common_mistakes?: string;
  keywords?: string;
  is_active: boolean;
  created_at: string;
  updated_at: string;
  related_concepts?: EducationalConceptSummary[];
}

export interface EducationalConceptSummary {
  id: number;
  name: string;
  concept_type: string;
  difficulty_level: string;
  short_description: string;
}

export interface ConceptExplanationRequest {
  concept_name: string;
  context?: string;
  user_level?: 'beginner' | 'intermediate' | 'advanced';
}

export interface ConceptExplanationResponse {
  concept: EducationalConcept;
  contextual_explanation: string;
  related_suggestions: EducationalConceptSummary[];
  next_learning_steps: string[];
}

export interface LearningProgress {
  id: number;
  user_id: number;
  concept_id: number;
  concept: EducationalConceptSummary;
  is_completed: boolean;
  completion_date?: string;
  difficulty_rating?: number;
  created_at: string;
  updated_at: string;
}

export interface LearningPathSuggestion {
  suggestions: string[];
  recommended_concepts: EducationalConceptSummary[];
  estimated_duration_minutes: number;
}

class EducationService {
  private baseUrl = '/api/v1/education';

  private getAuthHeaders(): HeadersInit {
    const token = localStorage.getItem('token');
    return {
      'Content-Type': 'application/json',
      ...(token && { 'Authorization': `Bearer ${token}` })
    };
  }

  /**
   * Search for educational concepts
   */
  async searchConcepts(
    query?: string,
    conceptType?: string,
    difficultyLevel?: string,
    limit: number = 20
  ): Promise<EducationalConcept[]> {
    const params = new URLSearchParams();
    if (query) params.append('query', query);
    if (conceptType) params.append('concept_type', conceptType);
    if (difficultyLevel) params.append('difficulty_level', difficultyLevel);
    params.append('limit', limit.toString());

    const response = await fetch(`${this.baseUrl}/concepts/search?${params}`, {
      headers: this.getAuthHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to search concepts: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get explanation for a specific concept
   */
  async explainConcept(request: ConceptExplanationRequest): Promise<ConceptExplanationResponse> {
    const response = await fetch(`${this.baseUrl}/explain`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify(request)
    });

    if (!response.ok) {
      throw new Error(`Failed to get concept explanation: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get user's learning progress
   */
  async getLearningProgress(): Promise<LearningProgress[]> {
    const response = await fetch(`${this.baseUrl}/progress`, {
      headers: this.getAuthHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to get learning progress: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Mark a concept as completed
   */
  async markConceptCompleted(
    conceptId: number,
    difficultyRating?: number
  ): Promise<LearningProgress> {
    const response = await fetch(`${this.baseUrl}/progress/${conceptId}`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify({
        completed: true,
        difficulty_rating: difficultyRating
      })
    });

    if (!response.ok) {
      throw new Error(`Failed to mark concept as completed: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get learning path suggestions
   */
  async getLearningPathSuggestions(
    userLevel: 'beginner' | 'intermediate' | 'advanced' = 'beginner'
  ): Promise<LearningPathSuggestion> {
    const response = await fetch(`${this.baseUrl}/learning-paths/suggestions?user_level=${userLevel}`, {
      headers: this.getAuthHeaders()
    });

    if (!response.ok) {
      throw new Error(`Failed to get learning path suggestions: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Get contextual learning suggestions based on chat content
   */
  async getContextualSuggestions(
    chatContent: string,
    stockSymbol?: string
  ): Promise<EducationalConceptSummary[]> {
    const response = await fetch(`${this.baseUrl}/contextual-suggestions`, {
      method: 'POST',
      headers: this.getAuthHeaders(),
      body: JSON.stringify({
        content: chatContent,
        stock_symbol: stockSymbol
      })
    });

    if (!response.ok) {
      throw new Error(`Failed to get contextual suggestions: ${response.statusText}`);
    }

    return response.json();
  }

  /**
   * Track concept interaction (for analytics)
   */
  async trackConceptInteraction(
    conceptName: string,
    interactionType: 'tooltip_view' | 'explanation_request' | 'completion',
    context?: string
  ): Promise<void> {
    try {
      await fetch(`${this.baseUrl}/track-interaction`, {
        method: 'POST',
        headers: this.getAuthHeaders(),
        body: JSON.stringify({
          concept_name: conceptName,
          interaction_type: interactionType,
          context: context
        })
      });
    } catch (error) {
      // Don't throw errors for tracking - it's not critical
      console.warn('Failed to track concept interaction:', error);
    }
  }
}

export const educationService = new EducationService();